src/routing/http-routing-reader-privacy-v1.md

```diff
@@ -0,0 +1,111 @@
+ - `SALT_ENCRYPTIONKEY = bytes("CR_ENCRYPTIONKEY\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00")`
+
+ Magic values are needed to calculate different digests from the same value for different purposes. For example a hash of a Multihash that is used for lookups should be different from the one that is used for
+ key derivation, even though both are calculated from the same original value. In order to do that the Multihash is concatenated with different magic values before applying the hash funciton - `SALT_DOUBLEHASH`
+ for lookups and `SALT_ENCRYPTIONKEY` for key derivation as described in the `Glossary`.
+
+### Glossary
+
+- **`enc`** is [AESGCM](https://en.wikipedia.org/wiki/Galois/Counter_Mode) encryption. The following notation will be used for the rest of the specification `enc(passphrase, nonce, payload)`.
+- **`hash`** is [SHA256](https://en.wikipedia.org/wiki/SHA-2) hashing.
+- **`||`** is concatenation of two values.
+- **`deriveKey`** is deriving a 32-byte encryption key from a passphrase that is done as `hash(SALT_ENCRYPTIONKEY || passphrase)`.
+- **`CID`** is the [Content IDentifier](https://github.com/multiformats/cid).
+- **`MH`** is the [Multihash](https://github.com/multiformats/multihash) contained in a `CID`. It corresponds to the 
+digest of a hash function over some content.
+- **`HASH2`** is a second hash over the multihash. Second Hashes must be of `Multihash` format with `DBL_SHA_256` codec. 
+The digest must be calculated as `hash(SALT_DOUBLEHASH || MH)`.
+- **`ProviderRecord`** is a JSON with Provider Record as described in the [HTTP Delegated Routing Specification](http-routing-v1.md).
+- **`ProviderRecordKey`** is a concatentation of `peerID || contextID`. There is no need for explicitly encoding lengths as they are
+already encoded as a part of the multihash format. Max `contextID` length is 64 bytes.
+- **`EncProviderRecordKey`** is `Nonce || enc(deriveKey(multihash), Nonce, ProviderRecordKey)`. Max `EncProviderRecordKey` is 200 bytes. 
+- **`HashProviderRecordKey`**  is a hash over `ProviderRecordKey` that must be calculated as `hash(SALT_DOUBLEHASH || ProviderRecordKey)`.
+- **`Metadata`** is free form bytes that can represent such information such as IPNI metadata. Max `Metadata` length is 1024 bytes. 
+- **`EncMetadata`** is `Nonce || enc(deriveKey(ProviderRecordKey), Nonce, Metadata)`. Max `EncMetadata` length is 2000 bytes.
+
+:::note
+
+Maximum allowed lengths might change without incrementing the API version. Such fields as `contextID` or `Metadata` are free-form bytes and
+their maximum lengths can be changed in the underlying protocols. 
+
+:::
+
+### API
+#### `GET /routing/v1/encrypted/providers/{HASH2}`
```

Comment:
Thinking out loud: @lidel I am curious what your thoughts are in terms of making this work with existing `/routing/v1/providers/{cid}`, where depending on the multihash code of `cid` in URL the router returns results in encrypted form. This is how the IPNI lookup API works and I think it offers a better UX, where there is a single endpoint for users to look things up at. Given a multihash the users would no longer have to swtich over different multihash codes to pick the endpoint.

Thoughts?

Reply:
@masih you're saying that if you see a CID where the multihash code is https://github.com/multiformats/multicodec/blob/b98f2f38fc63c992545263815c2afdd692d2dc8e/table.csv#L42 that you automatically assume it's a double-hashed record lookup? If so that's IMO quite a bad idea as it leads to domain mixing/confusion.

For example: `bafkvmibdoraoaissb2hgfuowe6kfpb3lrgxgfusza74f5i6zob6wl5p5uy` is a valid CID with data in it that can render on gateways (CAR attached for posterity [dblsha256-example.car.gz](https://github.com/ipfs/specs/files/11818033/dblsha256-example.car.gz)). However, with this approach it becomes impossible to discover this data via cid.contact.

---

Given the misunderstanding here I'm also concerned that IPNI has also misused the code as it's been defined since 2016 which is `SHA256(SHA256(data))` https://github.com/multiformats/multicodec/pull/23#discussion_r85876712 not say `SHA256(SHA256_MH(data)`, `SHA256("My Salt" | SHA256(data))`, etc. Am I missing something here?

Reply:
> ~you~ IPNI automatically assume(s) it's a double-hashed record lookup?

Some what. Backends that only store encrypted records refuse to store or lookup records with any other multihash. 
On loadbalancer, queries are scatter gathered across nodes. So if there is an unencrypted records associated to a double hashed multihash then it does get discovered. 

Having said that there are implicit assumptions in ipni to treat double hashed multicodec as encrypted record lookup.

> Am I missing something

Nope. Thank you for digging up those comments I was unaware that this code is from 2016. 
Looks like we cannot make that implicit assumption. Its reuse for reader privacy is maybe unfortunate as having a dedicated code would have help a bunch of optimisations in lookup path.